adapters/adkernelAdn/adkernelAdn.go

```diff
@@ -0,0 +1,242 @@
+	if imp.Banner == nil && imp.Video == nil {
+		return newBadInputError(fmt.Sprintf("Only banner and video impressions are supported. Ignoring imp id=%s", imp.ID))
+	}
+	if impExt.PublisherID < 1 {
+		return newBadInputError(fmt.Sprintf("Invalid pubId value. Ignoring imp id=%s", imp.ID))
+	}
+	return nil
+}
+
+//Group impressions by AdKernel-specific parameters `pubId` & `host`
+func dispatchImpressions(imps *[]openrtb.Imp, impsExt *[]openrtb_ext.ExtImpAdkernelAdn) map[openrtb_ext.ExtImpAdkernelAdn][]openrtb.Imp {
+	res := make(map[openrtb_ext.ExtImpAdkernelAdn][]openrtb.Imp)
+	for idx := range *imps {
+		imp := (*imps)[idx]
+		compatImpression(&imp)
+		impExt := (*impsExt)[idx]
+		if res[impExt] == nil {
+			res[impExt] = make([]openrtb.Imp, 0)
+		}
+		res[impExt] = append(res[impExt], imp)
+
+	}
+	return res
+}
+
+//Alter impression info to comply with adkernel platform requirements
+func compatImpression(imp *openrtb.Imp) error {
+	imp.Ext = nil //do not forward ext to adkernel platform
+	if imp.Banner != nil {
+		return compatBanerImpression(imp.Banner)
+	}
+	return nil
+}
+
+func compatBanerImpression(banner *openrtb.Banner) error {
+	//As banner.w/h are required fields for adkernel adn platform - take the first format entry
+	if banner.W == nil && banner.H == nil {
+		if len(banner.Format) == 0 {
+			return newBadInputError(fmt.Sprintf("Expected at least one banner.format entry"))
+		}
+		format := banner.Format[0]
+		banner.Format = banner.Format[1:]
+		banner.W = &format.W
+		banner.H = &format.H
+	}
+	return nil
+}
+
+func getImpressionExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpAdkernelAdn, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	var adkernelAdnExt openrtb_ext.ExtImpAdkernelAdn
+	if err := json.Unmarshal(bidderExt.Bidder, &adkernelAdnExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	return &adkernelAdnExt, nil
+}
+
+func buildAdapterRequest(prebidBidRequest *openrtb.BidRequest, params *openrtb_ext.ExtImpAdkernelAdn, imps *[]openrtb.Imp) (*adapters.RequestData, error) {
+	newBidRequest := createBidRequest(prebidBidRequest, params, imps)
+	reqJSON, err := json.Marshal(newBidRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     buildEndpointURL(params),
+		Body:    reqJSON,
+		Headers: headers}, nil
+}
+
+func createBidRequest(prebidBidRequest *openrtb.BidRequest, params *openrtb_ext.ExtImpAdkernelAdn, imps *[]openrtb.Imp) *openrtb.BidRequest {
+	bidRequest := *prebidBidRequest
+	bidRequest.Imp = *imps
+	for idx := range bidRequest.Imp {
+		imp := &bidRequest.Imp[idx]
+		imp.TagID = imp.ID
```

Comment:
This should result in `imp` being a copy of the imp in `bidRequest`, and then you set `TagID` in this copy. I don't see you using `imp` again, so I believe this won't do what you expect. You could try `bidRequest.Imp[idx].TagID = bidRequest.Imp[idx].ID`.

Reply:
@hhhjort My intention there was to write compact code but it looks like pointers are confusing. Regarding `bidRequest.Imp[idx].TagID = bidRequest.Imp[idx].ID` I have a noob question - does golang optimize double `bidRequest.Imp[idx]` op?

Reply:
That is a good question, I don't know for sure myself. Since one of the design goals of golang is to be a relatively simple language with just the minimum features needed, that allows for easier optimization in the compiler. Then again it is a relatively new language, so it is likely they haven't optimized every possible case.